a scenario on the stdlib and data types (#83)
digital-asset/daml-katacoda: fundamental-concepts/data-structures-stdlib/10_optional.md

<pre class="file" data-filename="daml/AddressBook.daml" data-target="append">
prettyEmail: Optional Text -> Text
prettyEmail mbEmail = case mbEmail of
  None -> " -- "
  Some email -> email
</pre>

Optional values can also be used as return values of functions, if you want to indicate that the
computation might fail or might be undefined. `lookup` of `DA.Next.Map` is an archetypal example:

<pre>
lookup: Map k v -> Optional v
</pre>

The `lookup` function makes it very clear by it's type that it might fail and there does not exist a
value with the given key. The `Optional` return value forces the consumer to deal with the possible
failure.

Have a look at the
[documentation](https://docs.daml.com/daml/stdlib/DA-Optional.html#module-da-optional), then try to
answer the following quiz:

>> Which functions converts an optional value to a list with possibly 0 or 1 element? <<
=== optionalToList

>> Which function of type f: a -> Optional a -> a converts an optional value to a value, given a default case for when the optional value is None ? <<
=== fromOptional

> >> Convert the list of optional values 'xs: [Optional Text]' to a list with of type '[Text]' which contains only the value constructed with 'Some'. <<
> === catOptionals xs
> 

Reviewer: ```suggestion
```
Would remove this question

Author: Wouldn't, `catOptionals` is quite an important and often used function.

Reviewer: Sure, but we can't teach them everything, some things they have to discover themselves. Biggest issue here is getting frustrated with the number of answers one needs to provide within the whole scenario (overall) as there's no option of progressing to the next step unless all the answers are correct. With previous questions on maps and lists you've really done a great job of pointing the users to the place where they can get necessary info on all things map related _when_ they decide to use it. That's the key.

> >> Take a look at 'DA.Either'. How could you get the same functionality as 'Optional a' with 'Either'? <<
> () Either Text a
> (*) Either () a
> () Either a a
> 
> `DA.Either` is another and richer way to encode errors in your data types. Instead of returning
> `None` in the error case, you can return `Left "a reason why the error happened"`. `DA.Either` is a
> good type to return for a function that can fail.

Reviewer: ```suggestion
We could get the same functionality as 'Optional a' with `DA.Either`, i.e., `Either () a`.
`DA.Either` is another and richer way to encode errors in your data types. Instead of returning
`None` in the error case, you can return `Left "a reason why the error happened"`. `DA.Either` is a
good type to return for a function that can fail.
```
I did look and unfortunately didn't understand form the docs how to use `Either` 😢 . Instead of asking the user to figure it out it might be better to tell her/him how to do it.

Author: removed it.
